### PR TITLE
Add unmapped command to read a summary of aria-details

### PR DESCRIPTION
**Pull request by seanbudd**
### Link to issue number:

None

### Summary of the issue:

`aria-details` is a HTML attribute that contains a reference to a HTML element that contains structured details. 

>aria-details is appropriate when linking to descriptions or annotations that are a bit more complex — rather than a simple text string, it might contain multiple bits of semantic information:

https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Annotations#aria-details_versus_aria-describedby

It would be helpful for NVDA users to be able to read a summary of `aria-details` before having to navigate through the structured annotation. 

### Description of how this pull request fixes the issue:

Adds an unmapped command to read a summary of `aria-details`. 

### Testing strategy:

System tests that read an example from chrome are implemented. 

### Known issues with pull request:

None

### Change log entries:

New Feature

- Adds a command that reads a summary of details marked for an object with `aria-details`. (#12364)

### Code Review Checklist:

- [x] Pull Request description is up to date.
- [x] Unit tests.
- [x] System (end to end) tests.
- [x] Manual tests.
- [x] User Documentation.
- [x] Change log entry.
- [x] Context sensitive help for GUI changes.

### CODE REVIEW

**Comment by XLTechie** [comment]
Are most users likely to even know that aria-details exist, or that they 
should access them?

Is there any indicator that aria-details are available?

**Comment by feerrenrut** [comment]
@XLTechie This is part of a much bigger block of work, which we intend to incrementally add support for.

File: source/browseMode.py [hunk 1]
```diff
@@ -1390,6 +1390,22 @@ def script_activateLongDesc(self,gesture):
...
 	script_activateLongDesc.__doc__=_("Shows the long description at this position if one is found.")
 
+	def script_activateAriaDetailsSummary(self, gesture):
+		info = self.makeTextInfo(textInfos.POSITION_CARET)
+		info.expand("character")
+		for field in reversed(info.getTextWithFields()):
+			if isinstance(field, textInfos.FieldCommand) and field.command == "controlStart":
+				states = field.field.get('states')
+				if states and controlTypes.STATE_HAS_ARIA_DETAILS in states:
+					speech.speakMessage(field.field['detailsSummary'])
```
Comment (feerrenrut) [inline source/browseMode.py:1400]: This should also output to braille. You can use the braille viewer to confirm. I think you can use `ui.message` for that.

File: source/controlTypes.py [hunk 2]
```diff
@@ -573,6 +574,8 @@
 	STATE_SORTED_DESCENDING:_("sorted descending"),
 	# Translators: a state that denotes that an object (usually a graphic) has a long description.
 	STATE_HASLONGDESC:_("has long description"),
+	# Translators: a state that denotes that an object has additional details (such as a comment section).
+	STATE_HAS_ARIA_DETAILS: _("has additional details"),
```
Comment (feerrenrut) [inline source/controlTypes.py:578]: I think we agreed on a shorter message: "`has details`"

**Comment by XLTechie** [comment]
Ah I see, thanks.

File: source/browseMode.py [hunk 3]
```diff
@@ -1398,6 +1398,22 @@ def script_activateLongDesc(self,gesture):
...
+				states = field.field.get('states')
+				if states and controlTypes.STATE_HAS_ARIA_DETAILS in states:
+					ui.message(field.field['detailsSummary'])
+					return
+
+		# Translators: the message presented when the activateAriaDetailsSummary script cannot locate a
+		# set of details to read.
+		ui.message(_("No additional details"))
+	# Translators: the description for the activateAriaDetailsSummary script on browseMode documents.
+	script_activateAriaDetailsSummary.__doc__ = _("Shows a summary of the details at this position if found.")
```
Comment (lukaszgo1) [inline source/browseMode.py:1415]: Please use script decorator here.

File: tests/system/robot/chromeTests.py [hunk 4]
```diff
@@ -48,6 +48,64 @@ def checkbox_labelled_by_inner_element():
...
+				</div>
+			</div>
+		</div>
+		"""
+	)
+	actualSpeech = _chrome.getSpeechAfterKey('downArrow')
+	# _asserts.strings_match(
+	# 	actualSpeech,
+	# 	"The word  marked content  has details  cat  out of marked content  has a comment tied to it."
+	# )
```
Comment (feerrenrut) [inline tests/system/robot/chromeTests.py:73]: Is this here for a reason? I assume just old code to be deleted?

File: tests/system/robot/chromeTests.py [hunk 5]
```diff
@@ -48,6 +48,64 @@ def checkbox_labelled_by_inner_element():
...
+	_asserts.strings_match(
+		actualSpeech,
+		"No additional details"
+	)
+	# this word has details attached to it
+	# actualSpeech = _chrome.getSpeechAfterKey("control+rightArrow")
+	# _asserts.strings_match(
+	# 	actualSpeech,
+	# 	"marked content  has details  cat  out of marked content"
+	# )
```
Comment (feerrenrut) [inline tests/system/robot/chromeTests.py:95]: Also?

**Review by feerrenrut — CHANGES_REQUESTED** [review]
Generally looks good. Take care with the C++ suggestions, please verify before relying on them, it seems easy to make mistakes with `BSTR`: https://docs.microsoft.com/en-us/cpp/atl/programming-with-ccombstr-atl?view=msvc-160

File: tests/system/robot/chromeTests.py (hunk 4 above)
Comment (seanbudd) [reply tests/system/robot/chromeTests.py]: yes this is to be introduced when we can introduce the discovery

File: nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp [hunk 6]
```diff
@@ -71,7 +79,10 @@ CComPtr<IAccessible2> GeckoVBufBackend_t::getLabelElement(IAccessible2_2* elemen
 	// we can now free the memory that Gecko  allocated to give us  the IUnknowns
 	CoTaskMemFree(ppUnk);
 	if(nTargets==0) {
-		LOG_DEBUG(L"relationTargetsOfType for IA2_RELATION_LABELLED_BY found no targets");
+		std::wstring debugMsg = L"relationTargetsOfType for ";
+		debugMsg.append((const wchar_t*) CComBSTR(ia2TargetRelation));
```
[the remark this answers is not in the document]
Comment (seanbudd) [reply nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp]: the simplest solution works

**Review by feerrenrut — CHANGES_REQUESTED** [review]

File: nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp [hunk 7]
```diff
@@ -63,8 +63,9 @@ CComPtr<IAccessible2> GeckoVBufBackend_t::getRelationElement(
 		numRelations=0;
 	}
 	// the relation type string *must* be passed correctly as a BSTR otherwise we can see crashes in 32 bit Firefox.
+	BSTR relationAsBSTR = CComBSTR(ia2TargetRelation);
```
[the remark this answers is not in the document]
Comment (feerrenrut) [reply nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp:66]: Looks like I made a typo: `relation` -> `relationAsBSTR`

**Review by feerrenrut — APPROVED** [review]
Code looks good. Assuming you have tested this in firefox and chrome again since your changes?

Lets talk about a merging strategy for this on Monday.

**Pull request merged**